controller/update.go

```diff
@@ -0,0 +1,255 @@
+package controller
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/giantswarm/inago/task"
+)
+
+type updateJobDesc struct {
+	Request Request
+}
+
+func (c controller) getNumRunningSlices(req Request) (int, error) {
+	unitStatusList, err := c.groupStatusWithValidate(req)
+	if err != nil {
+		return 0, maskAny(err)
+	}
+
+	var numRunning int
+	for _, us := range unitStatusList {
+		ok, err := unitHasStatus(us, StatusRunning)
+		if err != nil {
+			return 0, maskAny(err)
+		}
+		if ok {
+			numRunning++
+		}
+	}
+
+	return numRunning, nil
+}
+
+func (c controller) createJobQueues(req Request, opts UpdateOptions) (chan updateJobDesc, chan updateJobDesc, error) {
+	addQueue := make(chan updateJobDesc, len(req.SliceIDs))
+	removeQueue := make(chan updateJobDesc, len(req.SliceIDs))
+
+	numRunning, err := c.getNumRunningSlices(req)
+	if err != nil {
+		return nil, nil, maskAny(err)
+	}
+	numAllowedToRemove := numRunning - opts.MinAlive
+
+	var numRemovalQueued int
+	for _, sliceID := range req.SliceIDs {
+		jobDesc := updateJobDesc{
+			Request: Request{
+				Group:    req.Group,
+				SliceIDs: []string{sliceID},
+				Units:    req.Units,
+			},
+		}
+
+		addQueue <- jobDesc
+
+		if numAllowedToRemove >= numRemovalQueued {
+			removeQueue <- jobDesc
+			numRemovalQueued++
+		}
+	}
+
+	return addQueue, removeQueue, nil
+}
+
+func (c controller) isGroupRemovalAllowed(req Request, opts UpdateOptions) (bool, error) {
+	numRunning, err := c.getNumRunningSlices(req)
+	if err != nil {
+		return false, maskAny(err)
+	}
+
+	if numRunning > opts.MinAlive {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (c controller) isGroupAdditionAllowed(req Request, numTotal int, opts UpdateOptions) (bool, error) {
+	numRunning, err := c.getNumRunningSlices(req)
+	if err != nil {
+		return false, maskAny(err)
+	}
+
+	if numRunning < numTotal+opts.MaxGrowth {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (c controller) addWorker(jobDesc updateJobDesc, removeQueue chan updateJobDesc, fail chan<- error) {
```

Comment:
can we name this `runAddWorker`? I was confused because I assumed we add a worker here, not do the "add" steps